### PR TITLE
feat: replace and hidden options

### PR DESCRIPTION
**Pull request by rcarriga**
Adds the ability to replace existing notifications and omit
notifications from the history.

### CODE REVIEW

File: doc/nvim-notify.txt [hunk 1]
```diff
@@ -7,16 +7,28 @@ NotifyOptions                                                  *NotifyOptions*
...
-        {on_open}  (function|nil)  Callback for when window opens, receives
-                                   window as argument.
-        {on_close} (function|nil)  Callback for when window closes, receives
-                                   window as argument.
-        {keep}     (function|nil)  Function to keep the notification window
-                                   open after timeout, should return boolean.
-        {render}   (function)      Function to render a notification buffer.
+        {title}    (string)
+        {icon}     (string)
+        {timeout}  (number)                      Time to show notification in
```
Comment (vogelsgesang) [inline doc/nvim-notify.txt:12]: should probably mention `timeout=false` here

File: doc/nvim-notify.txt [hunk 2]
```diff
@@ -7,16 +7,28 @@ NotifyOptions                                                  *NotifyOptions*
...
+        {on_close} (function)                    Callback for when window
+                                                 closes, receives window as
+                                                 argument.
+        {keep}     (function)                    Function to keep the
+                                                 notification window open
+                                                 after timeout, should return
+                                                 boolean.
+        {render}   (function)                    Function to render a
+                                                 notification buffer.
+        {replace}  (integer|NotificationRecord)  Replace an existing
```
Comment (vogelsgesang) [inline doc/nvim-notify.txt:26]: what does `integer` do integers mean here? How do integers correspond to notifications?

File: doc/nvim-notify.txt [hunk 3]
```diff
@@ -7,16 +7,28 @@ NotifyOptions                                                  *NotifyOptions*
...
+                                                 notification if still open.
+                                                 All arguments not given are
+                                                 inherited from the replaced
+                                                 notification including
+                                                 message and level.
```
Comment (vogelsgesang) [inline doc/nvim-notify.txt:32]: `hidden` option seems to be missing from documentation

File: lua/notify/init.lua [hunk 4]
```diff
@@ -57,26 +57,32 @@ local function get_render(render)
...
----@field keep function | nil: Function to keep the notification window open after timeout, should return boolean.
+---@field title string
+---@field icon string
+---@field timeout number: Time to show notification in milliseconds.
+---@field on_open function: Callback for when window opens, receives window as argument.
+---@field on_close function: Callback for when window closes, receives window as argument.
+---@field keep function: Function to keep the notification window open after timeout, should return boolean.
 ---@field render function: Function to render a notification buffer.
+---@field replace integer | NotificationRecord: Replace an existing notification if still open. All arguments not given are inherited from the replaced notification including message and level.
+---@field hidden boolean Hide this notification from the history
```
Comment (vogelsgesang) [inline lua/notify/init.lua:68]: I would have expected the `hidden` flag to have the exact opposite effect: Hide the popup and only show the notification in the history which would be pretty useful for, e.g., debug messages, so I can still look them up in the history but am not spammed.

I also see value in "hide from history, only display". But given the potential for confusion, maybe we could find a more self-explaining name for this option?
E.g., `hide_from = "history"`? Could also be expanded to `hide_from = "popups"` or something similar in the future...

**Review by vogelsgesang — COMMENTED** [review]
Thanks for looking into this!

I will try it out next week. So far, I only read through the commit and left a couple of comments.

File: lua/notify/init.lua (hunk 4 above)
Comment (rcarriga) [reply lua/notify/init.lua]: Agreed the name is not very clear. I've gone with `hide_from_history` as a boolean option. I'll keep the potential for hiding notifications separate.

File: doc/nvim-notify.txt (hunk 2 above)
Comment (rcarriga) [reply doc/nvim-notify.txt]: Should be clear now :smile:

**Comment by will** [comment]
Is there an example anywhere of how to do the cool brail spinner?

**Comment by vogelsgesang** [comment]
See https://github.com/rcarriga/nvim-notify/wiki/Usage-Recipes

**Pull request merged**